src/cli.cpp

```diff
@@ -187,64 +126,146 @@ void THelpCmd::Usage() {
     if (ret) {
         PrintError("Unavailable");
     } else {
-        tmpVec.clear();
-        for (auto d : dlist)
-            tmpVec.push_back(d.Name);
-        nameWidth = MaxFieldLength(tmpVec);
+        nameWidth = MaxFieldLength(dlist, [](const TData &d) { return d.Name; });
 
-        for (auto d : dlist)
+        for (const auto& d : dlist)
             PrintAligned(d.Name, d.Description, nameWidth, termWidth);
     }
     std::cerr << std::endl;
 }
 
-int THelpCmd::Execute(int argc, char *argv[]) {
-    if (argc == 0) {
+int THelpCmd::Execute(TCommandEnviroment *env) {
+    int ret = EXIT_FAILURE;
+    const auto &args = env->GetArgs();
+
+    if (args.empty()) {
         Usage();
-        return EXIT_FAILURE;
+        return ret;
     }
 
-    string name(argv[0]);
-    for (auto i : commands) {
-        if (i.second->GetName() == name) {
-            std::cerr << "Usage: " << program_invocation_short_name << " " << name << " " << i.second->GetUsage() << std::endl;
-            std::cerr << std::endl;
-            std::cerr << i.second->GetDescription() << std::endl;
-            std::cerr << i.second->GetHelp();
-            return EXIT_SUCCESS;
-        }
+    const string &name = args[0];
+    const auto it = Handler.GetCommands().find(name);
+    if (it == Handler.GetCommands().end()) {
+        Usage();
+    } else {
+        std::cerr << "Usage: " << program_invocation_short_name
+                  << " " << name << " " << it->second->GetUsage()
+                  << std::endl;
+        std::cerr << std::endl;
+        std::cerr << it->second->GetDescription() << std::endl;
+        std::cerr << it->second->GetHelp();
+        ret = EXIT_SUCCESS;
     }
+    
+    return ret;
+}
+}  // namespace
 
-    Usage();
-    return EXIT_FAILURE;
+size_t MaxFieldLength(const std::vector<std::string> &vec, size_t min) {
+    return MaxFieldLength(vec, [](const string &s) { return s; }, min);
 }
 
-static void Usage(const char *command) {
-    ICmd *cmd = commands["help"];
-    char *argv[] = { (char *)command, NULL };
+int ICmd::RunCmdImpl(const std::vector<std::string> &args,
+                     std::unique_ptr<ICmd> cmd,
+                     TCommandEnviroment *env) {
+    ICmd *prevCmd = CurrentCmd;
+    cmd->SetDieOnSignal(false);
+
+    CurrentCmd = cmd.get();
+
+    TCommandEnviroment childEnv(env, args);
+    int ret = cmd->Execute(&childEnv);
+    CurrentCmd = prevCmd;
 
-    cmd->Execute(command ? 1 : 0, argv);
+    if (GotSignal())
+        return InterruptedSignal;
```

Comment:
During this refactoring I got some questions.
Could you give me some advices?
For example here we return just a `InterruptedSignal` value. But there is code line in TCommandHandler::TryExec where we return `-cmd->InterruptedSignal;` 
Is that OK?